# Agenda
- Pandas
    - Pandas in-built functions
    - Merging and Concatenation
    - Groupby with a case study
    - Saving and loading
- Doubt Clarification

# 2. Pandas in Python
- Pandas is an open-source Python Library providing high-performance data manipulation and analysis tool using its powerful data structures
- Using Pandas, we can accomplish five typical steps in the processing and analysis of data, regardless of the origin of data — load, prepare, manipulate, model, and analyze
- Python with Pandas is used in a wide range of fields including academic and commercial domains including finance, economics, Statistics, analytics, etc

### Key Features of Pandas
- Fast and efficient DataFrame object with default and customized indexing
- Tools for loading data into in-memory data objects from different file formats
- Data alignment and integrated handling of missing data
- Reshaping and pivoting of date sets
- Label-based slicing, indexing and subsetting of large data sets
- Columns from a data structure can be deleted or inserted
- Group by data for aggregation and transformations
- High performance merging and joining of data
- Time Series functionality

### Pandas deals with the following three data structures :
- Series
- DataFrame

These data structures are built on top of Numpy array, which means they are fast

### Mutability
- All Pandas data structures(series, dataframe) are value mutable (can be changed) and except Series all are size mutable. Series is size immutable

## 2.2 Pandas DataFrame
- DataFrame is a two-dimensional, size-mutable, potentially heterogeneous tabular data
- Data and Size both are mutable
- Labeled axes (rows and columns)
- Can Perform Arithmetic operations on rows and columns
- Potentially columns are of different types

## 2.5 Dataframe methods
- Similar to Series, Dataframe also has methods as :
  - describe()
  - count()
  - append()
  - apply()
  - columns()
  - dtypes()
  - astypes()
  - copy() and many more...

In [ ]:
import pandas as pd
dict = {'name':["John", "Puneet", "Sudhir", "Geeta"],
        'degree': ["MBA", "BCA", "M.Tech", "MBA"],
        'score':[90, 40, 80, 98]}
df = pd.DataFrame(dict)
df

In [ ]:
df

name  degree  score
0    John     MBA     90
1  Puneet     BCA     40
2  Sudhir  M.Tech     80
3   Geeta     MBA     98

#### describe()

In [ ]:
df.describe()

score
count   4.000000
mean   77.000000
std    25.742313
min    40.000000
25%    70.000000
50%    85.000000
75%    92.000000
max    98.000000

#### count()

In [ ]:
df.count()

name      4
degree    4
score     4
dtype: int64

#### append()

In [ ]:
df2 = pd.DataFrame([[1, 2], [3, 4]], columns=list('AB'), index=['x', 'y'])
df

A  B
x  1  2
y  3  4

In [ ]:
df3 = pd.DataFrame([[5, 6], [7, 8]], columns=list('AB'), index=['x', 'y'])
df3

A  B
x  5  6
y  7  8

In [ ]:
df2.append(df3)

A  B
x  1  2
y  3  4
x  5  6
y  7  8

#### apply()

In [ ]:
df2.apply(lambda x: x**2)

A   B
x  1   4
y  9  16

#### columns()

In [ ]:
df.columns

Index(['name', 'degree', 'score'], dtype='object')

#### dtypes()

In [ ]:
df.dtypes

name      object
degree    object
score      int64
dtype: object

#### astypes()

In [ ]:
df.astype({'score': 'float64'}).dtypes

name       object
degree     object
score     float64
dtype: object

#### copy()

In [ ]:
df2_copy = df2.copy()
df2_copy

A  B
x  1  2
y  3  4

## 2.6 Merging and Concatenating Dataframes
- **Need** - There are times when we need data from two or more dataframes in order to find relation between features. In such cases, pandas concat, merge, join functions are very handy and useful.

In [ ]:
df1 = pd.DataFrame(
    {
        "A": ["A0", "A1", "A2", "A3"],
        "B": ["B0", "B1", "B2", "B3"],
        "C": ["C0", "C1", "C2", "C3"],
        "D": ["D0", "D1", "D2", "D3"],
    },
    index=[0, 1, 2, 3],
)

df2 = pd.DataFrame(
    {
        "A": ["A4", "A5", "A6", "A7"],
        "B": ["B4", "B5", "B6", "B7"],
        "C": ["C4", "C5", "C6", "C7"],
        "D": ["D4", "D5", "D6", "D7"],
    },
    index=[4, 5, 6, 7],
)

In [ ]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [ ]:
df2

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

#### Concatenating dataframes
- Details of pandas concat function can be checked [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html)

In [ ]:
pd.concat([df1, df2])

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

Suppose we wanted to associate specific keys with each of the pieces of the chopped up DataFrame. We can do this using the keys argument.

In [ ]:
pd.concat([df1, df2], keys=["x", "y", "z"])

A   B   C   D
x 0  A0  B0  C0  D0
  1  A1  B1  C1  D1
  2  A2  B2  C2  D2
  3  A3  B3  C3  D3
y 4  A4  B4  C4  D4
  5  A5  B5  C5  D5
  6  A6  B6  C6  D6
  7  A7  B7  C7  D7

#### Merging Dataframes
- pandas merge function details can be checked [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html?highlight=merge#pandas.DataFrame.merge)

In [ ]:
df1 = pd.DataFrame({'lkey': ['foo', 'bar', 'baz', 'foo'],
                    'value': [1, 2, 3, 5]})
df2 = pd.DataFrame({'rkey': ['foo', 'bar', 'baz', 'foo'],
                    'value': [5, 6, 7, 8]})

In [ ]:
df1

lkey  value
0  foo      1
1  bar      2
2  baz      3
3  foo      5

In [ ]:
df2

rkey  value
0  foo      5
1  bar      6
2  baz      7
3  foo      8

In [ ]:
df1.merge(df2, left_on='lkey', right_on='rkey')

lkey  value_x rkey  value_y
0  foo        1  foo        5
1  foo        1  foo        8
2  foo        5  foo        5
3  foo        5  foo        8
4  bar        2  bar        6
5  baz        3  baz        7

## 2.7 Dataframe groupby method and Case Study

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
df = pd.read_csv("/content/drive/My Drive/nba.csv")
df.head()

Name            Team  Number  ... Weight            College     Salary
0  Avery Bradley  Boston Celtics     0.0  ...  180.0              Texas  7730337.0
1    Jae Crowder  Boston Celtics    99.0  ...  235.0          Marquette  6796117.0
2   John Holland  Boston Celtics    30.0  ...  205.0  Boston University        NaN
3    R.J. Hunter  Boston Celtics    28.0  ...  185.0      Georgia State  1148640.0
4  Jonas Jerebko  Boston Celtics     8.0  ...  231.0                NaN  5000000.0

[5 rows x 9 columns]

We can check information about this data and its dtypes using info() function

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


There are total 9 features in nba data. We will learn in detail about data pre processing such as categorical to numeric data conversion  and many more in upcoming weeks.
Aim of this section is to understand groupby function and its usage to make inferences

Lets check the shape of this data

In [ ]:
df.shape

(458, 9)

There are total 458 samples, lets check if there is any null value present. We can check it using pandas in built function called isna() or isnull()

In [ ]:
df.isna().sum()

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

###### There are some null values present in all the features, We can remove them. (How they can be removed will be covered in future sessions in detail)

In [ ]:
groups = df.groupby(by="Team")
groups

In [ ]:
# To see first values of each group
groups.first()

Name  ...      Salary
Team                                         ...            
Atlanta Hawks                 Kent Bazemore  ...   2000000.0
Boston Celtics                Avery Bradley  ...   7730337.0
Brooklyn Nets                  Markel Brown  ...    845059.0
Charlotte Hornets              Troy Daniels  ...    947276.0
Chicago Bulls              Cameron Bairstow  ...    845059.0
Cleveland Cavaliers     Matthew Dellavedova  ...   1147276.0
Dallas Mavericks            Justin Anderson  ...   1449000.0
Denver Nuggets               Darrell Arthur  ...   2814000.0
Detroit Pistons                Joel Anthony  ...   2500000.0
Golden State Warriors       Harrison Barnes  ...   3873398.0
Houston Rockets                Trevor Ariza  ...   8193030.0
Indiana Pacers                  Lavoy Allen  ...   4050000.0
Los Angeles Clippers           Cole Aldrich  ...   1100602.0
Los Angeles Lakers             Brandon Bass  ...   3000000.0
Memphis Grizzlies              Jordan Adams  ...   1404600.0
Miami Heat                       Chris Bosh  ...  22192730.0
Milwaukee Bucks              Jerryd Bayless  ...   3000000.0
Minnesota Timberwolves         Gorgui Dieng  ...   1474440.0
New Orleans Pelicans          Ryan Anderson  ...   8500000.0
New York Knicks               Arron Afflalo  ...   8000000.0
Oklahoma City Thunder          Steven Adams  ...   2279040.0
Orlando Magic                Dewayne Dedmon  ...    947276.0
Philadelphia 76ers            Isaiah Canaan  ...    947276.0
Phoenix Suns                   Eric Bledsoe  ...  13500000.0
Portland Trail Blazers      Cliff Alexander  ...    525093.0
Sacramento Kings                 Quincy Acy  ...    981348.0
San Antonio Spurs         LaMarcus Aldridge  ...  19689000.0
Toronto Raptors             DeMarre Carroll  ...  13600000.0
Utah Jazz                     Trevor Booker  ...   4775000.0
Washington Wizards            Alan Anderson  ...   4000000.0

[30 rows x 8 columns]

There are total 30 teams in nba data. Lets check data for "Washington Wizards" team of nba

In [ ]:
group_df.get_group("Washington Wizards")

Name  Number Position  ...  Weight               College      Salary
368    Alan Anderson     6.0       SG  ...   220.0        Michigan State   4000000.0
369     Bradley Beal     3.0       SG  ...   207.0               Florida   5694674.0
370     Jared Dudley     1.0       SF  ...   225.0        Boston College   4375000.0
371     Jarell Eddie     8.0       SG  ...   218.0         Virginia Tech    561716.0
372      Drew Gooden    90.0       PF  ...   250.0                Kansas   3300000.0
373    Marcin Gortat    13.0        C  ...   240.0                   NaN  11217391.0
374       JJ Hickson    21.0        C  ...   242.0  North Carolina State    273038.0
375     Nene Hilario    42.0        C  ...   250.0                   NaN  13000000.0
376  Markieff Morris     5.0       PF  ...   245.0                Kansas   8000000.0
377  Kelly Oubre Jr.    12.0       SF  ...   205.0                Kansas   1920240.0
378  Otto Porter Jr.    22.0       SF  ...   198.0            Georgetown   4662960.0
379   Ramon Sessions     7.0       PG  ...   190.0                Nevada   2170465.0
380   Garrett Temple    17.0       SG  ...   195.0                   LSU   1100602.0
381  Marcus Thornton    15.0       SF  ...   205.0                   LSU    200600.0
382        John Wall     2.0       PG  ...   195.0              Kentucky  15851950.0

[15 rows x 8 columns]

#### What is the average age of each team? (Hint : We will use an aggregate function along with groupby function)

In [ ]:
df.groupby("Team")[["Age"]].mean()

Age
Team                             
Atlanta Hawks           28.545455
Boston Celtics          23.833333
Brooklyn Nets           25.769231
Charlotte Hornets       25.692308
Chicago Bulls           27.333333
Cleveland Cavaliers     28.909091
Dallas Mavericks        29.000000
Denver Nuggets          27.777778
Detroit Pistons         26.200000
Golden State Warriors   26.583333
Houston Rockets         26.909091
Indiana Pacers          25.666667
Los Angeles Clippers    28.785714
Los Angeles Lakers      26.166667
Memphis Grizzlies       28.846154
Miami Heat              28.800000
Milwaukee Bucks         25.071429
Minnesota Timberwolves  24.000000
New Orleans Pelicans    26.437500
New York Knicks         27.000000
Oklahoma City Thunder   27.142857
Orlando Magic           25.200000
Philadelphia 76ers      23.714286
Phoenix Suns            25.000000
Portland Trail Blazers  25.066667
Sacramento Kings        26.538462
San Antonio Spurs       31.000000
Toronto Raptors         26.600000
Utah Jazz               24.600000
Washington Wizards      27.153846

#### We can also check average age and weight simultaneously for each team

In [ ]:
df.groupby("Team")[["Age", "Weight"]].mean()

Age      Weight
Team                                         
Atlanta Hawks           28.545455  220.181818
Boston Celtics          23.833333  218.000000
Brooklyn Nets           25.769231  216.153846
Charlotte Hornets       25.692308  216.692308
Chicago Bulls           27.333333  211.583333
Cleveland Cavaliers     28.909091  222.090909
Dallas Mavericks        29.000000  220.000000
Denver Nuggets          27.777778  208.666667
Detroit Pistons         26.200000  222.200000
Golden State Warriors   26.583333  225.833333
Houston Rockets         26.909091  213.909091
Indiana Pacers          25.666667  222.333333
Los Angeles Clippers    28.785714  222.214286
Los Angeles Lakers      26.166667  234.916667
Memphis Grizzlies       28.846154  219.153846
Miami Heat              28.800000  226.600000
Milwaukee Bucks         25.071429  222.642857
Minnesota Timberwolves  24.000000  217.500000
New Orleans Pelicans    26.437500  214.125000
New York Knicks         27.000000  223.636364
Oklahoma City Thunder   27.142857  228.285714
Orlando Magic           25.200000  216.000000
Philadelphia 76ers      23.714286  219.857143
Phoenix Suns            25.000000  215.153846
Portland Trail Blazers  25.066667  218.600000
Sacramento Kings        26.538462  221.923077
San Antonio Spurs       31.000000  220.818182
Toronto Raptors         26.600000  216.200000
Utah Jazz               24.600000  220.400000
Washington Wizards      27.153846  215.000000

#### What is the max salary given in each team?

In [ ]:
df.groupby("Team")["Salary"].max()

Team
Atlanta Hawks             18671659.0
Boston Celtics             7730337.0
Brooklyn Nets             19689000.0
Charlotte Hornets         12000000.0
Chicago Bulls             20093064.0
Cleveland Cavaliers       19689000.0
Dallas Mavericks          16407500.0
Denver Nuggets            11235955.0
Detroit Pistons           16000000.0
Golden State Warriors     15501000.0
Houston Rockets           15756438.0
Indiana Pacers            17120106.0
Los Angeles Clippers      21468695.0
Los Angeles Lakers        15592217.0
Memphis Grizzlies          9638555.0
Miami Heat                22192730.0
Milwaukee Bucks           16407500.0
Minnesota Timberwolves     5758680.0
New Orleans Pelicans      15514031.0
New York Knicks           22875000.0
Oklahoma City Thunder     20158622.0
Orlando Magic             11250000.0
Philadelphia 76ers         6500000.0
Phoenix Suns              13500000.0
Portland Trail Blazers     8042895.0
Sacramento Kings          15851950.0
San Antonio Spurs         1968900

## 2.8 Saving and Loading dataframe

#### csv file data loading and saving

### Data loading

In [ ]:
import pandas as pd
df = pd.read_csv("/content/drive/My Drive/nba.csv")
df.head()

Name            Team  Number  ... Weight            College     Salary
0  Avery Bradley  Boston Celtics     0.0  ...  180.0              Texas  7730337.0
1    Jae Crowder  Boston Celtics    99.0  ...  235.0          Marquette  6796117.0
2   John Holland  Boston Celtics    30.0  ...  205.0  Boston University        NaN
3    R.J. Hunter  Boston Celtics    28.0  ...  185.0      Georgia State  1148640.0
4  Jonas Jerebko  Boston Celtics     8.0  ...  231.0                NaN  5000000.0

[5 rows x 9 columns]

### Loading sklearn data

In [ ]:
from sklearn import datasets

In [ ]:
# we can load wine dataset from sklearn library
wine_data = load_wine()

In [ ]:
# we can load diabetes dataset from sklearn library
diabetes_data = load_diabetes()

### Loading Excel data

In [ ]:
import pandas as pd
df = pd.read_excel("/content/drive/My Drive/Product_List.xlsx")
df

Product  Price
0  Desktop Computer    700
1            Tablet    250
2           Printer    120
3            Laptop   1200

### Data saving

In [ ]:
df.to_csv("Week3_Profile_Data.csv")

### Pickle method

##### Do not unpickle data from untrusted source

data saving as pickle file on disk

In [ ]:
import pickle
df.to_pickle("Week3_Profile_Data.pkl")

data loading into csv file

In [ ]:
df1 = pd.read_pickle("Week3_Profile_Data.pkl")
df1

Unnamed: 0   Name  Age     Sex     City Country
0           0   Alex   30    Male       NY     USA
1           1  Anita   23  Female     Pune   India
2           2  Bliss   60  Female   Mumbai   India
3           3    Max   40    Male  Chicago     USA

## Happy Learning :)